tests/components/history_stats/test_sensor.py

```diff
@@ -1465,6 +1465,125 @@ def _fake_states(*args, **kwargs):
+            hass,
+            "sensor",
+            {
+                "sensor": [
+                    {
+                        "platform": "history_stats",
+                        "entity_id": "binary_sensor.state",
+                        "name": "sensor1",
+                        "state": "on",
+                        "start": "{{ today_at() }}",
+                        "end": "{{ now() }}",
+                        "type": "time",
+                    }
+                ]
+            },
+        )
+        await hass.async_block_till_done()
+
+        await async_update_entity(hass, "sensor.sensor1")
+        await hass.async_block_till_done()
+
+    assert hass.states.get("sensor.sensor1").state == "11.0"
+
+    # Advance 30 minutes
+    t1 = start_time + timedelta(minutes=30)
+    with freeze_time(t1):
+        async_fire_time_changed(hass, t1)
+        await hass.async_block_till_done()
+
+    assert hass.states.get("sensor.sensor1").state == "11.5"
+
+    # Advance 29 minutes, to record the last minute update just before midnight, just like a real system would do.
+    t2 = t1 + timedelta(minutes=29, microseconds=300)
+    with freeze_time(t2):
+        async_fire_time_changed(hass, t2)
+        await hass.async_block_till_done()
+
+    assert hass.states.get("sensor.sensor1").state == "11.98"
+
+    # One minute has passed and the time has now rolled over into a new day, resetting the recorder window. The sensor will then query the database for updates,
+    # and will see that the sensor is ON starting from midnight.
+    t3 = t2 + timedelta(minutes=1)
+
+    def _fake_states_t3(*args, **kwargs):
+        return {
+            "binary_sensor.state": [
+                ha.State(
+                    "binary_sensor.state",
+                    "on",
+                    last_changed=t3.replace(
+                        hours=0, minutes=0, seconds=0, microseconds=0
+                    ),
+                    last_updated=t3.replace(
+                        hours=0, minutes=0, seconds=0, microseconds=0
+                    ),
+                ),
+            ]
+        }
+
+    with (
+        patch(
+            "homeassistant.components.recorder.history.state_changes_during_period",
+            _fake_states_t3,
+        ),
+        freeze_time(t3),
+    ):
+        # The sensor turns off around this time, before the sensor does its normal polled update.
+        hass.states.async_set("binary_sensor.state", "off")
+        await hass.async_block_till_done(wait_background_tasks=True)
+
+    assert hass.states.get("sensor.sensor1").state == "0.0"
+
+    # More time passes, and the history stats does a polled update again. It should be 0 since the sensor has been off since midnight.
+    t4 = t3 + timedelta(minutes=10)
+    with freeze_time(t4):
+        async_fire_time_changed(hass, t4)
+        await hass.async_block_till_done()
+
+    assert hass.states.get("sensor.sensor1").state == "0.0"
```

Comment:
Without the fix the test fails here:

./tests/components/history_stats/test_sensor.py::test_state_change_during_window_rollover Failed: [undefined]AssertionError: assert '0.17' == '0.0'
  
  - 0.0
  + 0.17